archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,339 @@
+$\sum_{k=0}^{\infty}\frac{2}{2k+1}(\frac{1}{2a+1})^{2k+1}$.
+-/
+lemma power_series_log_succ_div (a : ℝ) (h : 0 < a) : has_sum (λ (k : ℕ),
+  (2 : ℝ) * (1 / (2 * (k : ℝ) + 1)) * ((1 / (2 * (a : ℝ) + 1)) ^ (2 * k + 1)))
+  (log ((a + 1) / a)) :=
+ begin
+  have h₁ : |1 / (2 * a + 1)| < 1, by --in library??
+  { rw [abs_of_pos, div_lt_one],
+    any_goals {linarith}, /- can not use brackets for single goal, bc of any_goals -/
+    {refine div_pos one_pos _, linarith, }, },
+  rw log_succ_div_eq_log_sub a (h),
+  exact log_sum_plus_minus (1 / (2 * a + 1)) h₁,
+ end
+
+/--
+`log_stirling_seq n` is log of `stirling_seq n`.
+-/
+noncomputable def log_stirling_seq (n : ℕ) : ℝ := log (stirling_seq n)
+
+/--
+We have the expression
+`log_stirling_seq (n + 1) = log(n + 1)! - 1 / 2 * log(2 * n) - n * log ((n + 1) / e)`.
+-/
+lemma log_stirling_seq_formula (n : ℕ): log_stirling_seq n.succ = (log (n.succ.factorial : ℝ)) -
+  1 / (2 : ℝ) * (log (2 * (n.succ : ℝ))) - (n.succ : ℝ) * log ((n.succ : ℝ) / (exp 1)) :=
+begin
+  have h3, from sqrt_ne_zero'.mpr (mul_pos two_pos $ cast_pos.mpr (succ_pos n)),
+  have h4 : 0 ≠ ((n.succ : ℝ) / exp 1) ^ n.succ, from
+    ne_of_lt (pow_pos (div_pos (cast_pos.mpr n.succ_pos ) (exp_pos 1)) n.succ),
+  rw [log_stirling_seq, stirling_seq, log_div, log_mul, sqrt_eq_rpow, log_rpow, log_pow],
+  { linarith },
+  { exact (zero_lt_mul_left zero_lt_two).mpr (cast_lt.mpr n.succ_pos),},
+  { exact h3, },
+  { exact h4.symm, },
+  { exact cast_ne_zero.mpr n.succ.factorial_ne_zero, },
+  { apply (mul_ne_zero h3 h4.symm), },
+end
+
+/--
+The sequence `log_stirling_seq (m + 1) - log_stirling_seq (m + 2)` has the series expansion
+   `∑ 1 / (2 * (k + 1) + 1) * (1 / 2 * (m + 1) + 1)^(2 * (k + 1))`
+-/
+lemma log_stirling_seq_diff_has_sum (m : ℕ) :
+  has_sum (λ (k : ℕ), (1 : ℝ) / (2 * k.succ + 1) * ((1 / (2 * m.succ + 1)) ^ 2) ^ (k.succ))
+  ((log_stirling_seq m.succ) - (log_stirling_seq m.succ.succ)) :=
+begin
+  change
+    has_sum ((λ (b : ℕ), 1 / (2 * (b : ℝ) + 1) * ((1 / (2 * (m.succ : ℝ) + 1)) ^ 2) ^ b) ∘ succ) _,
+  rw has_sum_nat_add_iff 1,
+  convert (power_series_log_succ_div m.succ (cast_pos.mpr (succ_pos m))).mul_left
+    ((m.succ : ℝ) + 1 / (2 : ℝ)),
+  { ext k,
+    rw [← pow_mul, pow_add],
+    have : 2 * (k : ℝ) + 1     ≠ 0, by {norm_cast, exact succ_ne_zero (2*k)},
+    have : 2 * (m.succ :ℝ) + 1 ≠ 0, by {norm_cast, exact succ_ne_zero (2*m.succ)},
+    field_simp,
+    ring },
+  { have h_reorder : ∀ {a b c d e f : ℝ}, a - 1 / (2 : ℝ) * b - c - (d - 1 / (2 : ℝ) * e - f) =
+      (a - d) - 1 / (2 : ℝ) * (b - e) - (c - f),
+    by {intros, ring_nf},
+    rw [log_stirling_seq_formula, log_stirling_seq_formula, h_reorder],
+    repeat {rw [log_div, factorial_succ]},
+    push_cast,
+    repeat {rw log_mul},
+    rw log_exp,
+    ring_nf,
+    all_goals {norm_cast},
+    all_goals {try {refine mul_ne_zero _ _}, try {exact succ_ne_zero _}},
```

Comment:
My personal suggestion is to avoid using these unless you have many very similar goals. Given how much these tactics are used, I doubt this is the case.

Reply:
At that point, I have 17 very similar goals open, I guess that counts as. The reason for this is the repeated application of the `log_div` lemma as well as the repeated `log_mul`. Both of those require two ne_zeros: `x ≠ 0` and `y ≠ 0`. 

Do you think I should as an alternative find out exactly what is `≠ 0` here, prove hypothesis for each of those and then use them explicitly when rewriting `log_mul` and `log_div`?

Reply:
If you have 17 different goals about things different to zero, this is fine. If they're all the same goal or two, you should just prove them individually.

Reply:
Hmm, I don't just mean this particular line, but the whole proof in lines 88-99.  I think there should be a way of writing it with a clearer structure. I don't have time to try it right now, I'll try to get to it tonight, but feel free to make your own experiments before then.

Reply:
Great, thanks! Perhaps @nick-kuhn can optimize this.

> If you have 17 different goals about things different to zero, this is fine. If they're all the same goal or two, you should just prove them individually.

The 17 goals are not all the same, but around 7 different ones.

Reply:
OK, I looked at this.  You can replace lines 85-99 with
```lean
simp only [log_stirling_seq_formula, log_div, log_mul, log_exp, factorial_succ, cast_mul,
  cast_succ, cast_zero, range_one, sum_singleton] { discharger :=
  `[norm_cast, apply_rules [mul_ne_zero, succ_ne_zero, factorial_ne_zero, exp_ne_zero]] },
ring
```

Reply:
Cool, done! Didn't know about `discharger`...